### ISSUE TITLE
[flang] false warning for C pointer to noninteroperable derived-type

### ISSUE THREAD

**Issue by ivan-pi** [llvm/llvm-project#167470]
I am trying to compile the following snippet which makes a pointer round-trip with a non-interoperable entity:

```fortran
program test
use, intrinsic :: iso_c_binding
implicit none

type :: foo
   integer :: b(3), c(2)
end type

type(foo), target :: x
type(foo), pointer :: p
type(c_ptr) :: pp

pp = c_loc(x)
call c_f_pointer(pp,p)

end program
```

This use case is important for interoperability, i.e. exposing Fortran classes and TPBs as opaque pointers in C. When applying `-pedantic` I see the warning:

```
/app/example.f90:14:21: portability: FPTR= argument to C_F_POINTER() should not have a derived type that is not BIND(C) [-Wportability]
  call c_f_pointer(pp,p)
````

I suspect this is a false positive; according to J3/24-007, pg 504, sec 18.2.3.3, the case above would fall under case (ii):

> If the value of CPTR is the result of a reference to C_LOC with a noninteroperable effective argument X, FPTR shall be a nonpolymorphic pointer with the same type and type parameters as X. In this case, X shall not have been deallocated or have become undefined due to execution of a RETURN or END statement since the reference.

A little variation of this example would be to use a polymorphic pointer argument:

```fortran
type(foo), target :: x
class(foo), pointer :: p
type(c_ptr) :: pp

pp = c_loc(x)
call c_f_pointer(pp,p) ! Error: p should not be polymorphic pointer
```

In this case, gfortran correctly reports: 

```
   14 | call c_f_pointer(pp,p)
      |                    1
Error: FPTR argument at (1) to C_F_POINTER shall not be polymorphic
```
Obviously, the compiler needs to know the true dynamic type (size, layout, vtable) to make the association. 

It would be of great value if flang could catch this class of errors.

**Comment by ivan-pi** [comment]
If anywhere, I guess a `class(smth), pointer` for case(iii) would fall under 19.5.3.2,

> In a storage association context, [...] (8) a data pointer occupies a single unspecified storage unit that is different from that of any nonpointer object and is different for each combination of type, type parameters, and rank. [...]

Cases (1)-(7) deal with nonpointer objects. Does storage sequence association extend to non-intrinsic types? 

Here is a slightly different example of why I think case(iii) does not apply here:

```fortran
use iso_c_binding

type :: smth
end type

class(smth), pointer :: x, px

allocate(smth :: x)  ! An object not in use by any other entity

! Argument to c_loc shall not be polymorphic
associate(ptr => c_loc(x))  ! Error

    ! ptr refers to object already in use by entity x, which
    ! is not in accord with case (iii)
    call c_f_pointer(ptr, px)
    
end associate
end
```

The other option would be something like this:

```fortran
interface
   ! Defined outside of Fortran
   function cptr_to_smth() bind(c)
      import c_ptr
      type(c_ptr) :: cptr_to_smth
   end function
end interface

class(smth), pointer :: px
call c_f_pointer(cptr_to_smth(), px)  ! case(iii)?
```

But how does one specify the correct size for a storage sequence of unspecified storage units? That seems tautological to me. Perhaps I'm missing something 🤷🏻‍♂️
